File: addons/mail/models/mail_message_reaction.py
import typing

from odoo import fields, models
from odoo.tools import groupby

from odoo.addons.mail.tools.discuss import Store, StoreFieldSpec

if typing.TYPE_CHECKING:
    from .discuss.mail_guest import MailGuest
    from .mail_message import MailMessage
    from .res_partner import ResPartner


class MailMessageReaction(models.Model):
    _name = "mail.message.reaction"
    _description = "Message Reaction"
    _order = "id desc"
    _log_access = False

    message_id: MailMessage = fields.Many2one(
        comodel_name="mail.message",
        index=True,
        readonly=True,
        required=True,
        ondelete="cascade",
    )
    content = fields.Char(
        readonly=True,
        required=True,
    )
    partner_id: ResPartner = fields.Many2one(
        comodel_name="res.partner",
        string="Reacting Partner",
        readonly=True,
        ondelete="cascade",
    )
    guest_id: MailGuest = fields.Many2one(
        comodel_name="mail.guest",
        string="Reacting Guest",
        readonly=True,
        ondelete="cascade",
    )

    _partner_unique = models.UniqueIndex(
        "(message_id, content, partner_id) WHERE partner_id IS NOT NULL"
    )
    _guest_unique = models.UniqueIndex(
        "(message_id, content, guest_id) WHERE guest_id IS NOT NULL"
    )

    _partner_or_guest_exists = models.Constraint(
        "CHECK((partner_id IS NOT NULL AND guest_id IS NULL) OR (partner_id IS NULL AND guest_id IS NOT NULL))",
        "A message reaction must be from a partner or from a guest.",
    )

    def _to_store(self, store: Store, fields: list[StoreFieldSpec]) -> None:
        if fields:
            raise NotImplementedError("Fields are not supported for reactions.")
        all_partner_ids = self.partner_id._ids
        all_guest_ids = self.guest_id._ids
        for (message, content), reactions in groupby(
            self, lambda r: (r.message_id, r.content)
        ):
            reactions = self.env["mail.message.reaction"].union(*reactions)
            data = {
                "content": content,
                "count": len(reactions),
                "guests": Store.Many(
                    reactions.guest_id.with_prefetch(all_guest_ids),
                    ["avatar_128", "name"],
                ),
                "message": message.id,
                "partners": Store.Many(
                    reactions.partner_id.with_prefetch(all_partner_ids),
                    ["avatar_128", *message._get_store_partner_name_fields()],
                ),
                "sequence": min(reactions.ids),
            }
            store.add_model_values("MessageReactions", data)

File: addons/mail/static/tests/mock_server/mock_models/mail_message_reaction.js
import { models } from "@web/../tests/web_test_helpers";

export class MailMessageReaction extends models.ServerModel {
    _name = "mail.message.reaction";
<<<<<<< a5db52fe916061e16bde55f411fa072b4e9105d9
||||||| 3b5eb0e1c36ef72c579f13f367f51c58a19205a1

    _to_store(store) {
        /** @type {import("mock_models").MailGuest} */
        const MailGuest = this.env["mail.guest"];
        /** @type {import("mock_models").ResPartner} */
        const ResPartner = this.env["res.partner"];

        const reactionGroups = groupBy(this, (r) => [r.message_id, r.content]);
        for (const groupId in reactionGroups) {
            const reactionGroup = reactionGroups[groupId];
            const { message_id, content } = reactionGroups[groupId][0];
            const guests = MailGuest.browse(reactionGroup.map((reaction) => reaction.guest_id));
            const partners = ResPartner.browse(
                reactionGroup.map((reaction) => reaction.partner_id)
            );
            const data = {
                content: content,
                count: reactionGroup.length,
                guests: mailDataHelpers.Store.many(
                    guests,
                    makeKwArgs({ fields: ["avatar_128", "name"] })
                ),
                message: message_id,
                partners: mailDataHelpers.Store.many(
                    partners,
                    makeKwArgs({ fields: ["avatar_128", "name"] })
                ),
                sequence: Math.min(reactionGroup.map((reaction) => reaction.id)),
            };
            store.add("MessageReactions", data);
        }
    }
=======

    _to_store(store) {
        /** @type {import("mock_models").MailGuest} */
        const MailGuest = this.env["mail.guest"];
        /** @type {import("mock_models").ResPartner} */
        const ResPartner = this.env["res.partner"];

        const reactionGroups = groupBy(this, (r) => [r.message_id, r.content]);
        for (const groupId in reactionGroups) {
            const reactionGroup = reactionGroups[groupId];
            const { message_id, content } = reactionGroups[groupId][0];
            const guests = MailGuest.browse(reactionGroup.map((reaction) => reaction.guest_id));
            const partners = ResPartner.browse(
                reactionGroup.map((reaction) => reaction.partner_id)
            );
            const data = {
                content: content,
                count: reactionGroup.length,
                guests: mailDataHelpers.Store.many(
                    guests,
                    makeKwArgs({ fields: ["avatar_128", "name"] })
                ),
                message: message_id,
                partners: mailDataHelpers.Store.many(
                    partners,
                    makeKwArgs({ fields: ["avatar_128", "name"] })
                ),
                sequence: Math.min(...reactionGroup.map((reaction) => reaction.id)),
            };
            store.add("MessageReactions", data);
        }
    }
>>>>>>> 98feb25ccf9378d6d9c383421e5f9cfe25dcd3cb
}
